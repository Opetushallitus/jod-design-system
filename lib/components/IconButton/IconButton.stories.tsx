import type { StoryObj } from '@storybook/react-vite';
import { fn } from 'storybook/test';
import type { TitledMeta } from '../../utils';
import { IconButton } from './IconButton';

const meta = {
  title: 'Buttons/IconButton',
  component: IconButton,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof IconButton>;

export default meta;

type Story = StoryObj<typeof meta>;

const Icon = () => {
  return (
    <svg
      stroke="currentColor"
      fill="currentColor"
      strokeWidth="0"
      viewBox="0 0 24 24"
      height="18px"
      width="18px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path fill="none" d="M0 0h24v24H0z"></path>
      <path d="M3 17v2h6v-2H3zM3 5v2h10V5H3zm10 16v-2h8v-2h-8v-2h-2v6h2zM7 9v2H3v2h4v2h2V9H7zm14 4v-2H11v2h10zm-6-4h2V7h4V5h-4V3h-2v6z"></path>
    </svg>
  );
};

const render = (args: Story['args']) => {
  return <IconButton {...args} icon={<Icon />} />;
};

export const Base: Story = {
  render,
  args: {
    label: 'Filter',
    onClick: fn(),
    icon: null,
  },
};
