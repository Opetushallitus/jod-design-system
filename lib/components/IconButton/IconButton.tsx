import { tidyClasses } from '../../utils';

export interface IconButtonProps {
  /** Text shown on the button */
  label: string;
  /** Hide label */
  hideLabel?: boolean;
  /** Callback fired on tap/click of the button */
  onClick: () => void;
  /** Button disabled for any actions */
  disabled?: boolean;
  /** Icon shown on the button. Icon size should be 18px */
  icon: React.ReactNode;
  /** Background color */
  bgColor?: 'gray' | 'white';
  dataTestId?: string;
}

export const IconButton = ({
  dataTestId,
  label,
  hideLabel = false,
  onClick,
  disabled = false,
  icon,
  bgColor = 'white',
}: IconButtonProps) => {
  const buttonBgColorClass = tidyClasses(`
    ${bgColor === 'gray' ? 'ds:bg-gray-500' : ''}
    ${bgColor === 'white' ? 'ds:bg-white' : ''}
  `);

  return (
    <button
      aria-label={label}
      type="button"
      onClick={onClick}
      disabled={disabled}
      data-testid={dataTestId}
      className={tidyClasses(`
        ${disabled ? 'ds:cursor-not-allowed ds:opacity-50' : 'ds:cursor-pointer'}
        ds:group
        ds:flex
        ds:flex-row
        ds:justify-center 
        ds:items-center
        ds:gap-4
      `)}
    >
      {!hideLabel && (
        <span
          className={tidyClasses(
            `ds:text-button-md ${disabled ? '' : 'ds:group-hover:text-accent ds:group-hover:underline'}`,
          )}
        >
          {label}
        </span>
      )}

      <div
        aria-hidden
        className={tidyClasses(`
            ${buttonBgColorClass}    
            ds:rounded-full
            ds:flex
            ds:items-center
            ds:justify-center
            ds:select-none
            ds:size-7
          `)}
      >
        {icon}
      </div>
    </button>
  );
};
