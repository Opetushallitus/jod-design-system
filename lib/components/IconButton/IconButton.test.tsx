import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, test, vi } from 'vitest';

import { IconButton } from './IconButton';

describe('Snapshot testing', () => {
  test('Default', () => {
    const { container } = render(<IconButton label="Default" onClick={vi.fn()} icon="target" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Disabled', () => {
    const { container } = render(<IconButton label="Disabled, non-selected" onClick={vi.fn()} icon="target" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

it('has the correct label', () => {
  render(<IconButton label="Label" onClick={vi.fn()} icon="target" />);
  expect(screen.getByRole('button', { name: 'Label' })).not.toBeNull();
});

it('calls the callback on click', async () => {
  const mockCallback = vi.fn();
  const user = userEvent.setup();

  render(<IconButton label="Test callback" onClick={mockCallback} icon="target" />);
  const button = screen.getByRole('button', { name: 'Test callback' });
  await user.click(button);
  expect(mockCallback).toHaveBeenCalledTimes(1);
});

it('does not call the callback while disabled', async () => {
  const mockCallback = vi.fn();
  const user = userEvent.setup();

  render(<IconButton label="Disabled IconButton" onClick={mockCallback} disabled icon="target" />);
  const button = screen.getByRole('button', { name: 'Disabled IconButton' });
  await user.click(button);
  expect(mockCallback).toHaveBeenCalledTimes(0);
});

describe('disabled', () => {
  it('should add correct CSS classes', () => {
    render(<IconButton label="Disabled IconButton" disabled={true} onClick={vi.fn()} icon="target" />);
    const parentClassList = screen.getByText('Disabled IconButton').parentElement?.classList;
    expect(parentClassList).toContain('ds:cursor-not-allowed');
    expect(parentClassList).toContain('ds:opacity-50');
  });

  it('has the disabled attribute', () => {
    const { getByRole } = render(
      <IconButton label="Disabled IconButton" disabled={true} onClick={vi.fn()} icon="target" />,
    );
    expect(getByRole('button')).toHaveProperty('disabled', true);
  });
});
