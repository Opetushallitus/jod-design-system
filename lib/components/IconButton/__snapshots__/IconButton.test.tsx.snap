// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > Default 1`] = `
<button
  aria-label="Default"
  class="ds:cursor-pointer ds:group ds:flex ds:flex-row ds:justify-center ds:items-center ds:gap-4"
  type="button"
>
  <span
    class="ds:text-button-md ds:group-hover:text-accent ds:group-hover:underline"
  >
    Default
  </span>
  <div
    aria-hidden="true"
    class="ds:bg-white ds:rounded-full ds:flex ds:items-center ds:justify-center ds:select-none ds:size-7"
  >
    target
  </div>
</button>
`;

exports[`Snapshot testing > Disabled 1`] = `
<button
  aria-label="Disabled, non-selected"
  class="ds:cursor-pointer ds:group ds:flex ds:flex-row ds:justify-center ds:items-center ds:gap-4"
  type="button"
>
  <span
    class="ds:text-button-md ds:group-hover:text-accent ds:group-hover:underline"
  >
    Disabled, non-selected
  </span>
  <div
    aria-hidden="true"
    class="ds:bg-white ds:rounded-full ds:flex ds:items-center ds:justify-center ds:select-none ds:size-7"
  >
    target
  </div>
</button>
`;
